### PR TITLE
Cluster capabilities

### PR DESCRIPTION
**Pull request by mrajashree**
https://github.com/rancher/rancher/issues/12422

types PR: https://github.com/rancher/types/pull/578

### CODE REVIEW

File: pkg/controllers/management/cluster/clustercontroller.go [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+	}
+	if cluster.Spec.RancherKubernetesEngineConfig != nil {
+		switch cluster.Spec.RancherKubernetesEngineConfig.CloudProvider.Name {
+		case aws.AWSCloudProviderName:
+			capabilities.L4LoadBalancer = c.L4Capability(true, "ELB", []string{"TCP"}, true)
+		case azure.AzureCloudProviderName:
+			capabilities.L4LoadBalancer = c.L4Capability(true, "Azure L4 LB", []string{"TCP", "UDP"}, true)
+		}
+		// only if not custom, non custom clusters have nodepools set
+		nodes, err := c.nodeLister.List(cluster.Name, labels.Everything())
```
Comment (alena1108) [inline pkg/controllers/management/cluster/clustercontroller.go]: should be enough to get just one node

File: pkg/controllers/management/cluster/clustercontroller.go [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+	c := controller{
+		clusterClient: management.Management.Clusters(""),
+		clusterLister: management.Management.Clusters("").Controller().Lister(),
+		nodeLister:    management.Management.Nodes("").Controller().Lister(),
+	}
+
+	c.clusterClient.AddHandler("clusterCreateUpdate", c.clusterSync)
+}
+
+func (c *controller) clusterSync(key string, obj *v3.Cluster) error {
```
Comment (alena1108) [inline pkg/controllers/management/cluster/clustercontroller.go]: can we rename the method to capsSync

File: pkg/controllers/management/cluster/clustercontroller.go [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+	var updated bool
+	if obj == nil || obj.DeletionTimestamp != nil {
+		return nil
+	}
+
+	cluster, err := c.clusterClient.Get(key, metav1.GetOptions{})
```
Comment (alena1108) [inline pkg/controllers/management/cluster/clustercontroller.go]: I don't think we should fetch the cluster at this point. We do have the cluster passed to the method, lets use that.

File: pkg/controllers/management/cluster/clustercontroller.go [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+	}
+
+	if k8serrors.IsNotFound(err) {
+		return nil
+	}
+
+	if cluster.Status.Driver == v3.ClusterDriverImported {
+		return nil
+	}
+	capabilities := v3.Capabilities{}
```
Comment (alena1108) [inline pkg/controllers/management/cluster/clustercontroller.go]: lets split populating various capabilities in its own submethod

File: pkg/controllers/management/cluster/clustercontroller.go [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+	capabilities.NodePortRange = "30000-32767"
+
+	if cluster.Spec.GoogleKubernetesEngineConfig != nil {
+		capabilities.L4LoadBalancer = c.L4Capability(true, "GCLB", []string{"TCP", "UDP"}, true)
```
Comment (alena1108) [inline pkg/controllers/management/cluster/clustercontroller.go]: lets think about making constants for the strings that are repeated several times

File: pkg/controllers/management/cluster/clustercontroller.go [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+		}
+
+		capabilities.IngressController = c.IngressCapability(true, "Nginx", false)
+		if cluster.Spec.RancherKubernetesEngineConfig.Services.KubeAPI.ExtraArgs["service-node-port-range"] != "" {
+			capabilities.NodePortRange = cluster.Spec.RancherKubernetesEngineConfig.Services.KubeAPI.ExtraArgs["service-node-port-range"]
+		}
+		updated = true
+	}
+
+	if updated {
```
Comment (alena1108) [inline pkg/controllers/management/cluster/clustercontroller.go]: we should determine the update purely by comparing capabilities field of existing cluster with newly generated capabilities

File: pkg/controllers/management/cluster/clustercontroller.go (hunk 4 above)
Comment (mrajashree) [reply pkg/controllers/management/cluster/clustercontroller.go]: I'm doing that for L4 and Ingress capability. NodePoolsScaling is just true/false and Node port range is also only one value so haven't done methods for that

File: pkg/controllers/management/cluster/clustercontroller.go (hunk 1 above)
Comment (mrajashree) [reply pkg/controllers/management/cluster/clustercontroller.go]: but I won't have the name of node

**Pull request merged**